src/com/amazon/corretto/crypto/provider/AesKeyWrapSpi.java

```diff
@@ -0,0 +1,227 @@
+                && !Arrays.equals(ICV2, ((IvParameterSpec) params).getIV())) {
+            throw new InvalidAlgorithmParameterException("Only ICV2 IvParameterSpec is accepted");
+        }
+        implInit(opmode, key, random);
+    }
+
+    private void implInit(int opmode, Key key, SecureRandom random)
+            throws InvalidKeyException, InvalidAlgorithmParameterException {
+        if (opmode != Cipher.UNWRAP_MODE && opmode != Cipher.WRAP_MODE) {
+            throw new UnsupportedOperationException("Cipher only supports un/wrap modes");
+        }
+        if (key == null) {
+            throw new InvalidKeyException("Null key");
+        }
+        if (key != jceKey) {
+            if (!(key instanceof  SecretKey)) {
+                throw new InvalidKeyException("Need a SecretKey");
+            }
+            if (!"RAW".equalsIgnoreCase(key.getFormat())) {
+                throw new InvalidKeyException("Need a raw format key");
+            }
+            if (!"AES".equalsIgnoreCase(key.getAlgorithm())) {
+                throw new InvalidKeyException("Expected an AES key");
+            }
+            if (keyBytes != null) {
+                Arrays.fill(keyBytes, (byte) 0);
+                keyBytes = null;
+            }
+            keyBytes = key.getEncoded();
+            if (keyBytes == null) {
+                throw new InvalidKeyException("Key doesn't support encoding");
+            }
+            jceKey = (SecretKey) key;
+        }
+        this.opmode = opmode;
+        // TODO [childw] do something with |random| here?
+    }
+
+    @Override
+    protected byte[] engineWrap(final Key key) throws IllegalBlockSizeException, InvalidKeyException {
+        if (opmode != Cipher.WRAP_MODE || keyBytes == null) {
+            throw new IllegalStateException("Cipher must be init'd in WRAP_MODE");
+        }
+        try {
+            final byte[] encoded = Utils.encodeForWrapping(provider, key);
+            final byte[] wrappedKey = new byte[4096];
+            int wrappedKeyLen = wrapKey(keyBytes, encoded, wrappedKey);
+            final byte[] wrappedKeyTruncated = new byte[wrappedKeyLen];
+            System.arraycopy(wrappedKey, 0, wrappedKeyTruncated, 0, wrappedKeyLen);
+            return wrappedKeyTruncated;
+        // TODO [childw] make this exception more specific, perhaps BadPaddingException?
+        } catch (final Exception ex) {
+            throw new InvalidKeyException("Wrapping failed", ex);
+        }
+    }
+
+    @Override
+    protected Key engineUnwrap(final byte[] wrappedKey, final String wrappedKeyAlgorithm, final int wrappedKeyType)
+            throws InvalidKeyException, NoSuchAlgorithmException {
+        if (opmode != Cipher.UNWRAP_MODE || keyBytes == null) {
+            throw new IllegalStateException("Cipher must be init'd in WRAP_MODE");
+        }
+        try {
+            final byte[] unwrappedKey = new byte[wrappedKey.length];
```

Comment:
You can estimate slightly tighter than this because you know that there is always at least 8 bytes of overhead (IIRC), so let's allocate the smaller amount.

Reply:
will do. unfortunately, i don't think we can guess the _precise_ size of the unwrapped key because of encrypted padding, [but we can get within 8 bytes](https://datatracker.ietf.org/doc/html/rfc5649#section-4.2). makes sense to minimize memory overhead.